# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: netwerk/protocol/http/HttpChannelParent.cpp
# Commit: 4f690ef38847
# Full Hash: 4f690ef388470e822103d1ae9a53aebc9e612e19
# Author: Paul Zuehlcke <pbz@mozilla.com>
# Date: 2023-09-19 03:53:26
# Regressor Bug: 1839918
# File Overlap Count: 2
# Description:
#   Bug 1839918 - Avoid returning IPC failures where possible in RecvSetCookies to avoid crashes. r=bvandersloot,necko-reviewers,jesup
#   
#   Depends on D186215
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D188329
# ==============================================================================

diff -r cc8f579c6e8e -r 4f690ef38847 netwerk/protocol/http/HttpChannelParent.cpp
--- a/netwerk/protocol/http/HttpChannelParent.cpp	Mon Sep 18 11:55:10 2023 +0000
+++ b/netwerk/protocol/http/HttpChannelParent.cpp	Mon Sep 18 11:55:10 2023 +0000
@@ -1047,9 +1047,7 @@
     nsIURI* aHost, const bool& aFromHttp, nsTArray<CookieStruct>&& aCookies) {
   net::PCookieServiceParent* csParent =
       LoneManagedOrNullAsserts(Manager()->ManagedPCookieServiceParent());
-  if (!csParent) {
-    return IPC_FAIL_NO_REASON(this);
-  }
+  NS_ENSURE_TRUE(csParent, IPC_OK());
 
   auto* cs = static_cast<net::CookieServiceParent*>(csParent);
 
@@ -1061,7 +1059,7 @@
   if (mBrowserParent) {
     dom::BrowsingContext* browsingContext =
         mBrowserParent->GetBrowsingContext();
-    if (browsingContext) {
+    if (browsingContext && !browsingContext->IsDiscarded()) {
       browsingContextId = browsingContext->Id();
       // Check if the cookies being set are third-party to the top level. We
       // do this by comparing the top level principals base domain with
@@ -1072,15 +1070,16 @@
 
         RefPtr<WindowGlobalParent> topWGP =
             topBC->Canonical()->GetEmbedderWindowGlobal();
-        NS_ENSURE_TRUE(topWGP, IPC_FAIL_NO_REASON(this));
+        if (!NS_WARN_IF(topWGP)) {
+          nsCOMPtr<nsIPrincipal> topPrincipal = topWGP->DocumentPrincipal();
+          MOZ_ASSERT(topPrincipal);
+          nsAutoCString topBaseDomain;
+          nsresult rv = topPrincipal->GetBaseDomain(topBaseDomain);
 
-        nsCOMPtr<nsIPrincipal> topPrincipal = topWGP->DocumentPrincipal();
-        MOZ_ASSERT(topPrincipal);
-        nsAutoCString topBaseDomain;
-        nsresult rv = topPrincipal->GetBaseDomain(topBaseDomain);
-        NS_ENSURE_SUCCESS(rv, IPC_FAIL_NO_REASON(this));
-
-        isThirdPartyCookie = aBaseDomain.Equals(topBaseDomain);
+          if (!NS_WARN_IF(NS_FAILED(rv))) {
+            isThirdPartyCookie = aBaseDomain.Equals(topBaseDomain);
+          }
+        }
       }
     }
   }
